File: NeetCode/Binary Search/Search2DMatrix.java
public class Search2DMatrix {
    public static boolean searchMatrix(int[][] matrix, int target) {
        int index = searchPotentialRow(matrix, target);        
        if(index != -1) return binarySearchOverRow(index, matrix, target);
        else return false;
    }

    public static int searchPotentialRow(int[][] matrix, int target) {
        int low = 0;
        int high = matrix.length-1;

        while(low <= high) {
            int middle = low + (high - low)/2;
            if(matrix[middle][0] <= target && target <= matrix[middle][matrix[0].length-1]) {
                return middle;
            } else if(target < matrix[middle][0]) {
                high = middle-1;
            } else if(target > matrix[middle][0]) {
                low = middle+1;
            }
        }
        
        return -1;
    }
    public static boolean binarySearchOverRow(int index, int[][] matrix, int target) {
        int low = 0;
        int high = matrix[index].length-1;

        while(low <= high) {
            int middle = low + (high - low)/2;
            if(target == matrix[index][middle]) {
                return true;
            } else if(target < matrix[index][middle]) {
                high = middle-1;
            } else if(target > matrix[index][middle]) {
                low = middle+1;
            }
        }

        return false;
    }
    
    public static void main(String[] args) {
        // int[][] matrix = {{1, 3, 5, 7}, {10, 11, 16, 20}, {23, 30, 34, 60}};
        // int target = 3;

        int[][] matrix = {{1, 3}};
        int target = 3;

        System.out.println(searchMatrix(matrix, target));
    }
}

File: NeetCode/Binary Search/Search2DMatrix.cs
public class Solution {
    public bool SearchMatrix(int[][] matrix, int target) {
        if (matrix == null || matrix.Length == 0 || matrix[0].Length == 0) {
            return false;
        } 
    
        int rows = matrix.Length;
        int elements = matrix[0].Length; 

        int row = SearchRow(matrix, target, 0, rows - 1);

        if (row != -1) {
            return SearchElement(matrix[row], target, 0, elements - 1);
        }

        return false;
    }

    protected int SearchRow(int[][] matrix, int target, int rowStart, int rowEnd) {
        while (rowStart <= rowEnd) {
            int mid = rowStart + (rowEnd - rowStart) / 2;
            int cols = matrix[0].Length;

            if (target >= matrix[mid][0] && target <= matrix[mid][cols - 1]) {
                return mid;
            } else if (target < matrix[mid][0]) {
                rowEnd = mid - 1;
            } else {
                rowStart = mid + 1;
            }
        }

        return -1;
    }

    protected bool SearchElement(int[] matrix, int target, int start, int end) {
        while (start <= end) {
            int mid = start + (end - start) / 2;

            if (target == matrix[mid]) {
                return true;
            } else if (target < matrix[mid]) {
                end = mid - 1;
            } else if (target > matrix[mid]) {
                start = mid + 1;
            }
        }

        return false;
    }
}
